src/ziplist.c

```diff
@@ -614,62 +614,99 @@ unsigned char *ziplistResize(unsigned char *zl, unsigned int len) {
+        if (cur.prevrawlen == prevlen) break;
 
-            /* Advance the cursor */
-            p += rawlen;
-            curlen += extra;
-        } else {
-            if (next.prevrawlensize > rawlensize) {
-                /* This would result in shrinking, which we want to avoid.
-                 * So, set "rawlen" in the available bytes. */
-                zipStorePrevEntryLengthLarge(p+rawlen,rawlen);
+        /* Abort when entry's "prevlensize" is big enough. */
+        if (cur.prevrawlensize >= prevlensize) {
+            if (cur.prevrawlensize == prevlensize) {
+                zipStorePrevEntryLength(p, prevlen);
             } else {
-                zipStorePrevEntryLength(p+rawlen,rawlen);
+                /* This would result in shrinking, which we want to avoid.
+                 * So, set "prevlen" in the available bytes. */
+                zipStorePrevEntryLengthLarge(p, prevlen);
             }
-
-            /* Stop here, as the raw length of "next" has not changed. */
             break;
         }
+
+        /* cur.prevrawlen means cur is the former head entry. */
+        assert(cur.prevrawlen == 0 || cur.prevrawlen + delta == prevlen);
+
+        /* Update prev entry's info and advance the cursor. */
+        rawlen = cur.headersize + cur.len;
+        prevlen = rawlen + delta; 
+        prevlensize = zipStorePrevEntryLength(NULL, prevlen);
+        prevoffset = p - zl;
+        p += rawlen;
+        extra += delta;
+        cnt++;
+    }
+
+    /* Extra bytes is zero all update has been done(or no need to update). */
+    if (extra == 0) return zl;
+
+    /* Update tail offset after loop. */
+    if (tail == zl + prevoffset) {
+        /* When the the last entry we need to update is also the tail, update tail offset
+         * unless this is the only entry that was updated (so the tail offset didn't change). */
+        if (extra - delta != 0) {
+            ZIPLIST_TAIL_OFFSET(zl) =
+                intrev32ifbe(intrev32ifbe(ZIPLIST_TAIL_OFFSET(zl))+extra-delta);
+        }
+    } else {
+        /* Update the tail offset in cases where the last entry we updated is not the tail. */
+        ZIPLIST_TAIL_OFFSET(zl) =
+            intrev32ifbe(intrev32ifbe(ZIPLIST_TAIL_OFFSET(zl))+extra);
+    }
+
+    /* Now "p" points at the first unchanged byte in original ziplist,
+     * move data after that to new ziplist. */
+    offset = p - zl;
+    zl = ziplistResize(zl, curlen + extra);
+    p = zl + offset;
+    memmove(p + extra, p, curlen - offset - 1);
+    p += extra;
+
+    /* Iterate all entries that need to be updated tail to head. */
+    while (cnt) {
+        zipEntry(zl + prevoffset, &cur);
+        rawlen = cur.headersize + cur.len;
+        /* Move entry to tail and reset prevlen. */
+        memmove(p - (rawlen - cur.prevrawlensize), 
+                zl + prevoffset + cur.prevrawlensize, 
+                rawlen - cur.prevrawlensize);
+        p -= (rawlen + delta);
+        if (cur.prevrawlen == 0) {
```

Comment:
thanks @neal-zhu , sorry I missed this edge cases.. the code logic is clear and looks good to me, also this is indeed a very good improvement.  Last thing is may I ask you did you run some functionaity tests for the cascade update? I think this is important to make sure we don't have any "surprise" in the future. If we can think about adding more tests in ziplist tests to cover this cascade update part, that would be nice..

Reply:
@hwware I had run the test of ziplist and it's all good. I did add a new test to benchmark this function when keeping inserting at the head of a ziplist which leads to cascade updating. 
Given that there is no semantics change, I think it may be enough to work with current tests?

Reply:
It doesn't look like the ziplist test is attempting to get to all the edge cases of the cascade (and other parts of the code).
E.g. First recorded being updated, first record being the only one that's updated, etc. 
I do think that there's a good chance that the redis test suite covers them, but it's hard to be certain.
I would feel better if you can add some tests, make sure they pass on both old and new implementations.

Reply:
@oranagra @hwware All right, I will add some tests for all the edge cases of the cascade update

Reply:
@neal-zhu thanks, I think the test it should looks like: 
1. create original ziplist which will cause cascade update if a insertion/delete or a merge operation happened.
2. doing insertion/delete or a merge operation.
3. check the ziplist fields (like the length, tail pos, entry field etc), make sure the new ziplist was updated correctly and no corruption.

Reply:
@oranagra @hwware Pls take a look at the new commit though it might not be thorough enough yet.

Reply:
thanks @neal-zhu for providing these tests, I loaded and test in my side did not see issue so far. Athough this is not 100% coverage for all the cases, maybe we can merge this into unstable branch and continusly test the behavior. At same time we can still improve the ziplist unit test.